### PR TITLE
Eliminate the Option from BufferedWriter::inner

### PR DESCRIPTION
**Pull request by reem**
The Option was used to allow moving out of self even though BufferedWriter
implements Drop, but the better solution is to just `ptr::read` the data
we need out, clean up the fields that need to be cleaned up, then simply
`mem::forget(self)`.

### CODE REVIEW

File: src/libstd/old_io/buffered.rs [hunk 1]
```diff
@@ -193,14 +195,14 @@ impl<W: Writer> BufferedWriter<W> {
...
 
     /// Gets a reference to the underlying writer.
-    pub fn get_ref(&self) -> &W { self.inner.as_ref().unwrap() }
+    pub fn get_ref(&self) -> &W { &self.inner }
 
     /// Gets a mutable reference to the underlying write.
     ///
     /// # Warning
     ///
-    /// It is inadvisable to directly read from the underlying writer.
```
Comment (huonw) [inline src/libstd/old_io/buffered.rs]: It's also inadvisable to read from it :P

**Comment by huonw** [comment]
I'm not super keen on the `unsafe` code, but there's not any other obvious choice. Although I do wonder how this fits into the new `io` lib.

r? @alexcrichton

**Comment by alexcrichton** [comment]
Out of curiosity, do you have benchmarks with before/after numbers showing that this is worth it? If so, can they be included in the test suite?

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to reopen with benchmarks!

**Pull request closed without merging**